### PR TITLE
Bluewire repo analysis tools

### PR DESCRIPTION
**Pull request by alex-davidson**
Implements the Bluewire.Tools.Runner tool host application, initially with only two children:
* `find-build`
* `generate-scripts`

### CODE REVIEW

File: Bluewire.Tools.GitRepository.IntegrationTests/BranchIntegrationPointLocatorTests.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+            await builder.AddCommitsToBranch("master", 2);
+
+            var resolved = await sut.FindCommit(workingCopy, thisStart, MasterBranch, new Ref("branch"));
+
+            Assert.That(resolved, Is.EqualTo(mergeCommit));
+        }
+
+        [Test, Description("Ancestry chains match entirely, but we know that subject != start: subject is the tip of the merged branch and was merged immediately after start.")]
+        public async Task SubjectAtTipOfBranchMergedImmediatelyAfterStartResolvesToIdentityOfMergeCommitImmediatelyFollowingStart()
```
Comment (luckyrat) [inline Bluewire.Tools.GitRepository.IntegrationTests/BranchIntegrationPointLocatorTests.cs]: Some of these mentions of "After start" and "Following start" are confusing me. Are they the same thing? If not, is there some alternative terminology we could use for one of them?

Comment (alex-davidson) [reply Bluewire.Tools.GitRepository.IntegrationTests/BranchIntegrationPointLocatorTests.cs]: I can't remember. I think there was a reason but it eludes me. I'll re-read it all.

Comment (alex-davidson) [reply Bluewire.Tools.GitRepository.IntegrationTests/BranchIntegrationPointLocatorTests.cs]: I think it has something to do with multiple commits capable of being 'after' the start commit, but only one can 'follow' it on the first-parent ancestry chain. Or something.

File: Bluewire.Tools.GitRepository.IntegrationTests/TopologicalBuildNumberResolverTests.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        {
+            await builder.CreateBranchWithCommits(startTag, "branch", 3);
+            await builder.AddCommitsToBranch("master", 8);
+
+            await session.Merge(workingCopy, new Ref("branch"));
+            var mergeCommit = await session.ResolveRef(workingCopy, MasterBranch);
+
+            await builder.AddCommitsToBranch("master", 3);
+
+
```
Comment (luckyrat) [inline Bluewire.Tools.GitRepository.IntegrationTests/TopologicalBuildNumberResolverTests.cs]: Too many blank lines.

File: Bluewire.Tools.GitRepository/TopologicalBuildNumberCalculator.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            if (workingCopyOrRepo == null) throw new ArgumentNullException(nameof(workingCopyOrRepo));
+            if (start == null) throw new ArgumentNullException(nameof(start));
+            if (subject == null) throw new ArgumentNullException(nameof(subject));
+
+            var cmd = session.CommandHelper.CreateCommand("rev-list", new Difference(start, subject), "--count");
+            var line = await session.CommandHelper.RunSingleLineCommand(workingCopyOrRepo, cmd);
+
+            int buildNumber;
+            if(!int.TryParse(line, out buildNumber)) return null;
```
Comment (luckyrat) [inline Bluewire.Tools.GitRepository/TopologicalBuildNumberCalculator.cs]: Missing space after if keyword.

File: Bluewire.Tools.GitRepository/TopologicalBuildNumberResolver.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            if (buildNumber == 0) return start;
+
+            var end = await session.ResolveRef(workingCopyOrRepo, endRef);
+            var endNumber = await calculator.GetBuildNumber(workingCopyOrRepo, start, end);
+            if (buildNumber == endNumber) return end;
+            if (buildNumber > endNumber) throw new BuildNumberOutOfRangeException($"Build number {buildNumber} exceeds the last known build number: {endNumber}");
+
+            var candidatesDescending = await session.ListCommitsBetween(workingCopyOrRepo, start, end, new ListCommitsOptions { FirstParentOnly = true });
+
+            var target =  await SearchImplementation.Search(candidatesDescending, buildNumber, r => calculator.GetBuildNumber(workingCopyOrRepo, start, r));
```
Comment (luckyrat) [inline Bluewire.Tools.GitRepository/TopologicalBuildNumberResolver.cs]: extra space after =

File: Bluewire.Conventions/BranchSemantics.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+namespace Bluewire.Conventions
+{
+    public class BranchSemantics
+    {
+        public BranchType GetBranchType(StructuredBranch branch)
+        {
+            if (branch.Name == "master") return BranchType.Beta;
+
+            var lastNamespacePart = branch.Namespace?.Substring(branch.Namespace.LastIndexOf('/') + 1);
+            if(lastNamespacePart == null) return BranchType.None;
```
Comment (luckyrat) [inline Bluewire.Conventions/BranchSemantics.cs]: missing space after `if`

File: Bluewire.Conventions/SprintNumber.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        /// Expects a two-part version number with no whitespace.
+        /// </summary>
+        /// <returns>Null if it's not a sprint number.</returns>
+        public static Version Parse(string str)
+        {
+            if (str.Any(Char.IsWhiteSpace)) return null;
+            Version version;
+            if (!Version.TryParse(str, out version)) return null;
+            // Check that we only got the first two components:
+            if(version.Revision < 0 && version.Build < 0) return version;
```
Comment (luckyrat) [inline Bluewire.Conventions/SprintNumber.cs]: Missing space after `if`

File: Bluewire.Tools.Runner/FindBuild/RepositoryStructureInspector.cs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        {
+            // Optimisation: only have Git explore the branches we care about.
+            var branchFilter = new BranchSemantics().GetRemoteBranchFilters(types.Concat(new [] { BranchType.Master }).ToArray());
+
+            var containingBranches = await gitSession.ListBranches(repository, new ListBranchesOptions { Contains = hash, Remote = true, BranchFilter = branchFilter });
+
+            StructuredBranch? masterBranch;
+            var recognisedBranches = ParseRecognisedBranches(types, containingBranches, out masterBranch);
+            if (recognisedBranches.Length <= 1) return recognisedBranches;
+            if (recognisedBranches.Length <= pruneThreshold) return recognisedBranches;
```
Comment (luckyrat) [inline Bluewire.Tools.Runner/FindBuild/RepositoryStructureInspector.cs]: Looks like we only need one of the above conditionals. Maybe put a check at the top of the function to ensure pruneThreshold is always >= 1 instead?

File: Bluewire.Tools.Runner/FindBuild/ResolveBuildVersionsFromGitHubPullRequest.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        private async Task<Ref> ResolveToSingleCommit(GitSession session, Common.GitWrapper.GitRepository repository)
+        {
+            var pattern = $"^Merge pull request #{pullRequestNumber}\\b";
+
+            var prMerges = await session.ReadLog(repository, new LogOptions { MatchMessage = new Regex(pattern), IncludeAllRefs = true });
+
+            if (prMerges.Length == 1) return prMerges.Single().Ref;
+            if (prMerges.Length == 0) throw new ErrorWithReturnCodeException(3, $"Could not find a merge of PR #{pullRequestNumber}.");
+
+            var mergedCommits = prMerges.Select(m => m.MergeParents?.Skip(1)).Aggregate((rs, r) => r == null ? Enumerable.Empty<Ref>() : rs.Intersect(r)).ToArray();
```
Comment (luckyrat) [inline Bluewire.Tools.Runner/FindBuild/ResolveBuildVersionsFromGitHubPullRequest.cs]: This `Skip(1)` is potentially a little opaque so perhaps include a quick comment to explain it's related to the git merge parent conventions/specification?

File: Bluewire.Tools.Runner/FindBuild/ResolveBuildVersionsFromTicketIdentifier.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        public async Task<string[]> ResolveBuildVersions(GitSession session, Common.GitWrapper.GitRepository repository)
+        {
+            var hashes = await ResolveToHashes(session, repository);
+
+            // Tracing target branches can be expensive, so we should try to cull the hash list first.
+            // If A is an ancestor of B, then the set of branches containing A must be a superset of the set of
+            // branches containing B (since all branches containing B must contain A as well).
+            // But --is-ancestor is rather expensive too, especially for an O(n^2) operation...
```
Comment (luckyrat) [inline Bluewire.Tools.Runner/FindBuild/ResolveBuildVersionsFromTicketIdentifier.cs]: This confused me a bit. I think you're talking about the implementation details of some of the methods that are called by this method but I'm not sure exactly how it relates to this particular method. Is it essentially aiming to explain why we are calling `IsChildOfAny` on each hash? It looks like that calls `--is-ancestor` so I'm not sure what to make of the comment that says this operation is expensive. I assume it's been found to be faster than not culling the list of hashes at all?

File: Bluewire.Tools.Runner/FindBuild/TargetBranchResolver.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            if (gitSession == null) throw new ArgumentNullException(nameof(gitSession));
+            if (repository == null) throw new ArgumentNullException(nameof(repository));
+            this.gitSession = gitSession;
+            this.repository = repository;
+        }
+
+        public async Task<StructuredBranch[]> IdentifyTargetBranchesOfCommit(Ref hash)
+        {
+            var inspector = new RepositoryStructureInspector(gitSession);
```
Comment (luckyrat) [inline Bluewire.Tools.Runner/FindBuild/TargetBranchResolver.cs:24]: Do we need to create this object every time we call this method? It looks like it is tied to the `gitSession` which is not going to change so why not make this an instance property and lazily instantiate the backing field instead? Alternatively, if there is some state within the inspector that we want to discard every time we invoke `IdentifyTargetBranchesOfCommit`, perhaps making the inspector disposable and wrapping its use in a using block would make this more explicit.

File: Bluewire.Tools.Runner/GenerateScripts/ToolRunner.cs [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+            {
+                return $@"#!/bin/sh
+{shellRunnerInvocation} --tool {toolName} ""$@""
+";
+            }
+            private string GetCommandFileContent(string runnerInvocation, string toolName)
+            {
+                return $@"@echo off
+{runnerInvocation} --tool {toolName} %*
+";
```
Comment (luckyrat) [inline Bluewire.Tools.Runner/GenerateScripts/ToolRunner.cs:100]: There are some weird line endings in the strings returned from this and the above method. Is that intentional? If so, why are only 3 of the 4 line endings represented in that way?

**Review by luckyrat — CHANGES_REQUESTED** [review]
Looks great apart from those few issues I've mentioned, mostly minor whitespace stuff. If you could look into the naming of those tests too that would be good. Maybe once you've understood their purpose, we can brainstorm a clearer name for them.

File: Bluewire.Tools.Runner/FindBuild/TargetBranchResolver.cs (hunk 10 above)
Comment (alex-davidson) [reply Bluewire.Tools.Runner/FindBuild/TargetBranchResolver.cs:24]: It's a service object. It doesn't store any state and merely encapsulates logic. There's no particular need to keep it around between calls.

File: Bluewire.Tools.Runner/FindBuild/RepositoryStructureInspector.cs (hunk 7 above)
Comment (alex-davidson) [reply Bluewire.Tools.Runner/FindBuild/RepositoryStructureInspector.cs]: I'll reorder the conditionals to make it clearer why I did it like this.

File: Bluewire.Tools.Runner/GenerateScripts/ToolRunner.cs (hunk 11 above)
Comment (alex-davidson) [reply Bluewire.Tools.Runner/GenerateScripts/ToolRunner.cs:100]: This is a GitHub rendering artefact. Examining the file shows that all line breaks are CRLF.

**Comment by alex-davidson** [comment]
~~Build failure on TeamCity.~~ Intermittent failure.

**Review by luckyrat — APPROVED** [review]

**Pull request merged**